File: packages/core/src/schemas.ts
import { z } from "zod";
import {
  AGGRESSION_LEVELS,
  APPROVAL_ACTIONS,
  ASSET_TYPES,
  BROLL_SOURCES,
  CAMERA_STYLES,
  CAMPAIGN_SOURCE_TYPES,
  CAMPAIGN_STATUSES,
  CAMPAIGN_TYPES,
  CONTENT_TYPES,
  DIRECTNESS_LEVELS,
  MEMORY_TYPES,
  PERSONA_STATUSES,
  PERSONA_TYPES,
  PLATFORMS,
  PLUG_FREQUENCIES,
  PLUG_TYPES,
  RISK_LEVELS,
} from "./enums.js";

// ---------------------------------------------------------------------------
// API input schemas
// ---------------------------------------------------------------------------

export const personaCreateSchema = z.object({
  name: z.string().min(1).max(120),
  slug: z
    .string()
    .min(1)
    .max(120)
    .regex(/^[a-z0-9-]+$/)
    .optional(),
  personaType: z.enum(PERSONA_TYPES).default("CUSTOM"),
  status: z.enum(PERSONA_STATUSES).default("ACTIVE"),
  description: z.string().max(2000).default(""),
  backstory: z.string().max(20000).default(""),
  worldview: z.string().max(5000).default(""),
  speakingStyle: z.string().max(5000).default(""),
  tone: z.string().max(1000).default(""),
  humorStyle: z.string().max(1000).default(""),
  // Required: every persona must carry disclosure language (§0.1).
  disclosureText: z.string().min(1).max(1000),
  defaultLanguage: z.string().max(20).default("en"),
  defaultPlatformTone: z.string().max(2000).default(""),
  memoryEnabled: z.boolean().default(true),
});
export type PersonaCreateInput = z.infer<typeof personaCreateSchema>;

export const personaUpdateSchema = personaCreateSchema.partial();

export const personaMemoryCreateSchema = z.object({
  type: z.enum(MEMORY_TYPES),
  title: z.string().min(1).max(300),
  content: z.string().min(1).max(20000),
  importance: z.number().min(0).max(1).default(0.5),
  source: z.string().max(300).default("manual"),
});

export const personaAssetCreateSchema = z.object({
  type: z.enum(ASSET_TYPES),
  name: z.string().min(1).max(300),
  url: z.string().url(),
  s3Key: z.string().max(1000).optional(),
  metadata: z.record(z.string(), z.unknown()).default({}),
  isDefault: z.boolean().default(false),
});

export const campaignCreateSchema = z.object({
  personaId: z.string().min(1),
  name: z.string().min(1).max(200),
  slug: z
    .string()
    .min(1)
    .max(200)
    .regex(/^[a-z0-9-]+$/)
    .optional(),
  status: z.enum(CAMPAIGN_STATUSES).default("ACTIVE"),
  campaignType: z.enum(CAMPAIGN_TYPES).default("CUSTOM"),
  objective: z.string().max(5000).default(""),
  targetAudience: z.string().max(2000).default(""),
  productName: z.string().max(300).default(""),
  productDescription: z.string().max(10000).default(""),
  productUrl: z.string().url().optional(),
  directnessLevel: z.enum(DIRECTNESS_LEVELS).default("CASUAL"),
  plugFrequency: z.enum(PLUG_FREQUENCIES).default("WHEN_NATURAL"),
  plugPercentage: z.number().int().min(0).max(100).optional(),
  primaryKpi: z.string().max(100).default("VIEWS"),
  secondaryKpis: z.array(z.string().max(100)).default([]),
  startDate: z.coerce.date().optional(),
  endDate: z.coerce.date().optional(),
});
export type CampaignCreateInput = z.infer<typeof campaignCreateSchema>;

export const campaignUpdateSchema = campaignCreateSchema
  .omit({ personaId: true })
  .partial();

export const campaignSourceCreateSchema = z.object({
  type: z.enum(CAMPAIGN_SOURCE_TYPES),
  title: z.string().min(1).max(300),
  content: z.string().max(100000).default(""),
  url: z.string().url().optional(),
  fileUrl: z.string().url().optional(),
});

export const guardrailConfigSchema = z.object({
  allowedTopics: z.array(z.string().max(300)).default([]),
  bannedTopics: z.array(z.string().max(300)).default([]),
  allowedClaims: z.array(z.string().max(1000)).default([]),
  bannedClaims: z.array(z.string().max(1000)).default([]),
  requiredDisclosures: z.array(z.string().max(1000)).default([]),
  competitorRules: z
    .object({
      allowCompetitorMentions: z.boolean().default(false),
      competitorNames: z.array(z.string().max(200)).default([]),
    })
    .default({ allowCompetitorMentions: false, competitorNames: [] }),
  aggressionLevel: z.enum(AGGRESSION_LEVELS).default("NORMAL"),
  sensitiveTopicRules: z.array(z.string().max(1000)).default([]),
  escalationRules: z.array(z.string().max(1000)).default([]),
  platformSpecificRules: z.record(z.string(), z.array(z.string())).default({}),
  wordsToAvoid: z.array(z.string().max(100)).default([]),
});
export type GuardrailConfigInput = z.infer<typeof guardrailConfigSchema>;

export const generationRequestSchema = z.object({
  campaignId: z.string().min(1),
  contentType: z.enum(CONTENT_TYPES).default("TEXT_POST"),
  platform: z.enum(PLATFORMS).default("X"),
  instruction: z.string().max(5000).optional(),
  /** The comment being replied to — required for REPLY / DM_REPLY. */
  originalComment: z.string().max(10000).optional(),
});
export type GenerationRequest = z.infer<typeof generationRequestSchema>;

export const approvalActionSchema = z.object({
  action: z.enum(APPROVAL_ACTIONS),
  reason: z.string().max(5000).optional(),
  editInstruction: z.string().max(5000).optional(),
  /** Target platform for CHANGE_PLATFORM regenerations. */
  platform: z.enum(PLATFORMS).optional(),
  edits: z
    .object({
      title: z.string().max(300).optional(),
      hook: z.string().max(1000).optional(),
      script: z.string().max(20000).optional(),
      caption: z.string().max(5000).optional(),
      bodyText: z.string().max(20000).optional(),
      hashtags: z.array(z.string().max(100)).optional(),
      cta: z.string().max(1000).optional(),
      scheduledFor: z.coerce.date().optional(),
    })
    .optional(),
});
export type ApprovalActionInput = z.infer<typeof approvalActionSchema>;

// ---------------------------------------------------------------------------
// LLM structured-output schemas (plan §8)
// ---------------------------------------------------------------------------

export const brollPlanItemSchema = z.object({
  timestamp: z.string().describe('e.g. "0:03-0:06"'),
  description: z.string(),
  source: z.enum(BROLL_SOURCES),
});

export const generatedShortVideoSchema = z.object({
  title: z.string(),
  hook: z.string(),
  script: z.string(),
  caption: z.string(),
  hashtags: z.array(z.string()),
  cta: z.string(),
  visualDirection: z.string(),
  cameraStyle: z.enum(CAMERA_STYLES),
  outfitSuggestion: z.string(),
  brollPlan: z.array(brollPlanItemSchema),
  thumbnailIdea: z.string(),
  campaignPlugType: z.enum(PLUG_TYPES),
  whyThisShouldWork: z.string(),
  riskNotes: z.array(z.string()),
});
export type GeneratedShortVideo = z.infer<typeof generatedShortVideoSchema>;

export const generatedTextPostSchema = z.object({
  body: z.string(),
  hook: z.string(),
  cta: z.string(),
  hashtags: z.array(z.string()),
  campaignPlugType: z.enum(PLUG_TYPES),
  whyThisShouldWork: z.string(),
  riskNotes: z.array(z.string()),
});
export type GeneratedTextPost = z.infer<typeof generatedTextPostSchema>;

export const generatedReplySchema = z.object({
  reply: z.string(),
  tone: z.string(),
  shouldMentionCampaign: z.boolean(),
  campaignMention: z.string(),
  riskLevel: z.enum(RISK_LEVELS),
  escalateToHuman: z.boolean(),
  reason: z.string(),
});
export type GeneratedReply = z.infer<typeof generatedReplySchema>;

// ---------------------------------------------------------------------------
// Guardrail result (plan §7.9)
// ---------------------------------------------------------------------------

export const guardrailResultSchema = z.object({
  passed: z.boolean(),
  riskScore: z.number().min(0).max(1),
  warnings: z.array(z.string()),
  blockers: z.array(z.string()),
  requiredEdits: z.array(z.string()),
});
export type GuardrailResult = z.infer<typeof guardrailResultSchema>;

File: apps/pipeline/hype_pipeline/schemas.py
"""Pydantic models for the pipeline service.

These mirror the shared shapes in packages/core (zod). Content-plan models
use extra='forbid' so their JSON schema carries additionalProperties: false,
as required by Anthropic structured outputs. Keep fields constraint-free
(no min/max/pattern) — structured outputs reject those; semantic checks live
in validators or the guardrail engine instead.
"""
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, field_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


# ---------- LLM output plans (mirror packages/core content-schemas) ----------

BrollSource = Literal["AI_GENERATED", "USER_ASSET", "SCREEN_RECORDING", "STOCK"]
PlugType = Literal["NONE", "SUBTLE", "CASUAL", "DIRECT"]
RiskLevel = Literal["LOW", "MEDIUM", "HIGH"]


class BrollShot(StrictModel):
    timestamp: str
    description: str
    source: BrollSource


class ShortVideoPlan(StrictModel):
    title: str
    hook: str
    script: str
    caption: str
    hashtags: list[str]
    cta: str
    visualDirection: str
    cameraStyle: str
    outfitSuggestion: str
    brollPlan: list[BrollShot]
    thumbnailIdea: str
    campaignPlugType: PlugType
    whyThisShouldWork: str
    riskNotes: list[str]
    sourceCitations: list[str]


class TextPostPlan(StrictModel):
    body: str
    hook: str
    cta: str
    hashtags: list[str]
    campaignPlugType: PlugType
    whyThisShouldWork: str
    riskNotes: list[str]
    sourceCitations: list[str]


class ReplyPlan(StrictModel):
    reply: str
    tone: str
    shouldMentionCampaign: bool
    campaignMention: str
    riskLevel: RiskLevel
    escalateToHuman: bool
    reason: str


class LearningInsightPlan(StrictModel):
    insight: str
    evidence: str
    confidence: float
    actionRecommendation: str

    @field_validator("confidence")
    @classmethod
    def _clamp_confidence(cls, v: float) -> float:
        return max(0.0, min(1.0, v))


# ---------- Guardrails ----------

class GuardrailPolicy(BaseModel):
    campaignType: str
    allowedTopics: list[str] = []
    bannedTopics: list[str] = []
    allowedClaims: list[str] = []
    bannedClaims: list[str] = []
    requiredDisclosures: list[str] = []
    wordsToAvoid: list[str] = []
    allowCompetitorMentions: bool = False
    competitorNames: list[str] = []
    directnessLevel: str = "CASUAL"


class ChecklistItem(BaseModel):
    label: str
    passed: bool
    detail: Optional[str] = None


class GuardrailResult(BaseModel):
    passed: bool
    riskScore: int
    riskLevel: RiskLevel
    warnings: list[str]
    blockers: list[str]
    requiredEdits: list[str]
    checklist: list[ChecklistItem]


# ---------- API contract ----------

class PersonaContext(BaseModel):
    name: str
    backstory: str = ""
    worldview: str = ""
    speakingStyle: str = ""
    tone: str = ""
    humorStyle: str = ""
    disclosureText: str
    memoryHighlights: list[dict[str, str]] = []


class CampaignSource(BaseModel):
    type: str
    title: str
    content: str = ""


class CampaignContext(BaseModel):
    name: str
    campaignType: str
    objective: str = ""
    subject: Optional[str] = None
    productName: Optional[str] = None
    productDescription: Optional[str] = None
    productUrl: Optional[str] = None
    targetAudience: Optional[str] = None
    directnessLevel: str = "CASUAL"
    plugFrequency: str = "WHEN_NATURAL"
    plugPercentage: Optional[int] = None
    mainMessage: Optional[str] = None
    productLine: Optional[str] = None
    sources: list[CampaignSource] = []
    learnings: list[str] = []


OutputKind = Literal[
    "SHORT_VIDEO", "TEXT_POST", "THREAD", "REPLY", "DM_REPLY", "WHATSAPP_MESSAGE"
]


class GenerateRequest(BaseModel):
    outputKind: OutputKind
    platform: str
    contentType: str
    persona: PersonaContext
    campaign: CampaignContext
    policy: GuardrailPolicy
    includePlug: Optional[bool] = None
    extraInstructions: Optional[str] = None
    recentSummaries: list[str] = []
    recentTexts: list[str] = []


class ContentFields(BaseModel):
    title: Optional[str] = None
    hook: Optional[str] = None
    script: Optional[str] = None
    caption: Optional[str] = None
    bodyText: Optional[str] = None
    hashtags: list[str] = []
    cta: Optional[str] = None


class Usage(BaseModel):
    inputTokens: int = 0
    outputTokens: int = 0
    costUsd: float = 0.0


class GenerateResponse(BaseModel):
    fields: ContentFields
    plugType: PlugType
    riskNotes: list[str]
    sourceCitations: list[str]
    metadata: dict
    usage: Usage
    guardrails: GuardrailResult
    generationPrompt: str


class EvaluateRequest(BaseModel):
    policy: GuardrailPolicy
    platform: str
    contentType: str = ""
    title: Optional[str] = None
    hook: Optional[str] = None
    script: Optional[str] = None
    caption: Optional[str] = None
    bodyText: Optional[str] = None
    cta: Optional[str] = None
    hashtags: list[str] = []
    campaignPlugType: PlugType = "NONE"
    riskNotes: list[str] = []
    sourceCitations: list[str] = []
    recentTexts: list[str] = []


class InsightsRequest(BaseModel):
    persona: PersonaContext
    campaignName: str
    objective: str
    optimizationTarget: str = "REACH"
    performanceLines: list[str]


class InsightsResponse(BaseModel):
    insight: LearningInsightPlan
    usage: Usage
